presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeMatchingChecker.java

```diff
@@ -0,0 +1,169 @@
+
+            checkExpressionTypeMatching(node.getAssignments());
+
+            return null;
+        }
+
+        @Override
+        public Void visitUnion(UnionNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            checkMappingTypeMatching(node.getSymbolMapping());
+
+            return null;
+        }
+
+        private void checkMappingTypeMatching(ListMultimap<Symbol, Symbol> symbolMapping)
+        {
+            for (Symbol keySymbol : symbolMapping.keySet()) {
+                List<Symbol> valueSymbols = symbolMapping.get(keySymbol);
+                Type type = types.get(keySymbol);
+                for (Symbol valueSymbol : valueSymbols) {
+                    verifyTypeSignature(keySymbol, type.getTypeSignature(), types.get(valueSymbol).getTypeSignature());
+                }
+            }
+        }
+
+        private void checkSignatureTypeMatching(Map<Symbol, Signature> functions)
+        {
+            for (Map.Entry<Symbol, Signature> entry : functions.entrySet()) {
+                TypeSignature typeSignature = types.get(entry.getKey()).getTypeSignature();
+                TypeSignature functionTypeSignature = entry.getValue().getReturnType();
+                verifyTypeSignature(entry.getKey(), typeSignature, functionTypeSignature);
+            }
+        }
+
+        private void checkFunctionCallTypeMatching(Map<Symbol, FunctionCall> functionCalls)
+        {
+            for (Map.Entry<Symbol, FunctionCall> entry : functionCalls.entrySet()) {
+                Type type = types.get(entry.getKey());
+                Type expressionType = getExpressionTypes(session, metadata, sqlParser, types, entry.getValue()).get(entry.getValue());
+                verifyTypeSignature(entry.getKey(), type.getTypeSignature(), expressionType.getTypeSignature());
+            }
+        }
+
+        private void checkExpressionTypeMatching(Map<Symbol, Expression> expressions)
+        {
+            for (Map.Entry<Symbol, Expression> entry : expressions.entrySet()) {
+                Type type = types.get(entry.getKey());
+                if (entry.getValue() instanceof QualifiedNameReference) {
+                    QualifiedNameReference qualifiedNameReference = (QualifiedNameReference) entry.getValue();
+                    verifyTypeSignature(entry.getKey(), type.getTypeSignature(), types.get(Symbol.fromQualifiedName(qualifiedNameReference.getName())).getTypeSignature());
+                    continue;
+                }
+                Type expressionType = getExpressionTypes(session, metadata, sqlParser, types, entry.getValue()).get(entry.getValue());
+                verifyTypeSignature(entry.getKey(), type.getTypeSignature(), expressionType.getTypeSignature());
+            }
+        }
+
+        private static void verifyTypeSignature(Symbol symbol, TypeSignature expected, TypeSignature actual)
+        {
+            if (!actual.equals(UNKNOWN.getTypeSignature())) {
```

Comment:
The special case is that when we assign `null` to a typed symbol, `null` will be of `UNKNOWN`. So here we should consider `UNKNOWN` as a wildcard, which match any other types.

I will also add a comment to explain it.